### PR TITLE
Prefer `Arrays.copyOf` over `new Array` + `System.arraycopy`

### PR DESCRIPTION
**Pull request by armanbilge**
`Arrays.copyOf` is an intrinsic.

### CODE REVIEW

**Review by durban — APPROVED** [review]

**Comment by djspiewak** [comment]
This looks very very interesting. I definitely wanna run some benchmarks.

**Comment by He-Pin** [comment]
I think should be the nearly the same, and the jit compiler will inline anyway, maybe the old one is more performance?

File: core/jvm-native/src/main/scala/cats/effect/ByteStack.scala [hunk 1]
```diff
@@ -45,9 +47,7 @@ private object ByteStack {
     if ((1 + ((count + 1) >> 3)) < stack.length) {
       stack
     } else {
-      val bigger = new Array[Int](stack.length << 1)
-      System.arraycopy(stack, 0, bigger, 0, stack.length) // Count in stack(0) copied "for free"
-      bigger
+      Arrays.copyOf(stack, 2 * stack.length) // Count in stack(0) copied "for free"
```
Comment (He-Pin) [inline core/jvm-native/src/main/scala/cats/effect/ByteStack.scala:50]: (stack.length << 1) be faster or the compiler already did that.

Comment (djspiewak) [reply core/jvm-native/src/main/scala/cats/effect/ByteStack.scala:50]: In theory the jit should have been doing that for us.

**Comment by djspiewak** [comment]
> I think should be the nearly the same, and the jit compiler will inline anyway

Well, if *all* of `copyOf` is intrinsic compared to just the `arraycopy` operation itself, then it should theoretically be faster by some amount. It's tricky though because this isn't really the bottleneck, so it might not show up in benchmarks. I'm going to find out though.

File: core/jvm-native/src/main/scala/cats/effect/ByteStack.scala (hunk 1 above)
Comment (He-Pin) [reply core/jvm-native/src/main/scala/cats/effect/ByteStack.scala:50]: That's true,but it took time for the jit to generate that code.

**Comment by He-Pin** [comment]
Do you have any number @armanbilge eg with a jmh？I want to do the same in pekko too, if you like ,you can do that too. May be there are some special handling in Hotspot.

**Comment by He-Pin** [comment]
@IntrinsicCandidate，cool

**Comment by He-Pin** [comment]
@glavo is doing

**Comment by He-Pin** [comment]
- [ ] She just share the number

**Comment by Glavo** [comment]
Testing platform:

```
> screenfetch
                          ./+o+-       glavo@minecraft-server
                  yyyyy- -yyyyyy+      OS: Ubuntu 22.04 jammy
               ://+//////-yyyyyyo      Kernel: x86_64 Linux 6.2.0-37-generic
           .++ .:/++++++/-.+sss/`      Uptime: 53d 10h 7m
         .:++o:  /++++++++/:--:/-      Packages: 2270
        o:+o+:++.`..```.-/oo+++++/     Shell: fish 3.3.1
       .:+o:+o/.          `+sssoo+/    Disk: 360G / 938G (41%)
  .++/+:+oo+o:`             /sssooo.   CPU: AMD Ryzen 7 5800X 8-Core @ 16x 4.6GHz
 /+++//+:`oo+o               /::--:.   GPU: GeForce GT 710
 \+/+o+++`o++o               ++////.   RAM: 12780MiB / 31995MiB
  .++.o+++oo+:`             /dddhhh.
       .+.o+oo:.          `oddhhhh+
        \+.++o+o``-````.:ohdhhhhh+
         `:o+++ `ohhhhhhhhyo++os:
           .o:`.syhhhhhhh/.oo++o`
               /osyyyyyyo++ooo+++/
                   ````` +oo+++o\:
                          `oo++.
> java -version
openjdk version "21.0.1" 2023-10-17 LTS
OpenJDK Runtime Environment (build 21.0.1+12-LTS)
OpenJDK 64-Bit Server VM (build 21.0.1+12-LTS, mixed mode, sharing)
```

Benchmark:

```java
import org.openjdk.jmh.annotations.*;

import java.util.Arrays;
import java.util.concurrent.TimeUnit;

@Warmup(iterations = 5, time = 3)
@Measurement(iterations = 5, time = 3)
@Fork(value = 1, jvmArgsAppend = {"-XX:+UseG1GC", "-Xms8g", "-Xmx8g", "--add-exports=java.base/jdk.internal.misc=ALL-UNNAMED", "-Xint"})
@BenchmarkMode(Mode.Throughput)
@OutputTimeUnit(TimeUnit.MILLISECONDS)
@State(Scope.Benchmark)
public class ArrayCopy {
    @Param({"1", "4", "16", "64", "256", "1024", "4096"})
    public int length;

    private Object[] array;

    @Setup
    public void setup() {
        array = new Object[length];
        for (int i = 0; i < length; i++) {
            array[i] = new Object();
        }
    }

    @Benchmark
    public Object[] arraycopy() {
        Object[] res = new Object[length * 2];
        System.arraycopy(array, 0, res, 0, length);
        return res;
    }

    @Benchmark
    public Object[] arraysCopyOf() {
        return Arrays.copyOf(array, length * 2);
    }

}
```

JMH results:

```
Benchmark               (length)   Mode  Cnt     Score      Error   Units
ArrayCopy.arraycopy            1  thrpt    5  7775.857 ± 1053.637  ops/ms
ArrayCopy.arraycopy            4  thrpt    5  8010.345 ± 1093.602  ops/ms
ArrayCopy.arraycopy           16  thrpt    5  8183.028 ±  376.640  ops/ms
ArrayCopy.arraycopy           64  thrpt    5  7812.816 ±  484.204  ops/ms
ArrayCopy.arraycopy          256  thrpt    5  4699.294 ±  150.074  ops/ms
ArrayCopy.arraycopy         1024  thrpt    5  1472.525 ±   52.211  ops/ms
ArrayCopy.arraycopy         4096  thrpt    5   394.825 ±    0.393  ops/ms
ArrayCopy.arraysCopyOf         1  thrpt    5  4314.874 ±  145.806  ops/ms
ArrayCopy.arraysCopyOf         4  thrpt    5  4115.145 ±  243.908  ops/ms
ArrayCopy.arraysCopyOf        16  thrpt    5  4493.738 ±  213.516  ops/ms
ArrayCopy.arraysCopyOf        64  thrpt    5  4326.075 ±  139.554  ops/ms
ArrayCopy.arraysCopyOf       256  thrpt    5  2966.048 ±   48.140  ops/ms
ArrayCopy.arraysCopyOf      1024  thrpt    5  1255.927 ±    6.378  ops/ms
ArrayCopy.arraysCopyOf      4096  thrpt    5   377.148 ±    0.573  ops/ms
```

Counter-intuitively, `Arrays.copyOf` is slower (especially for small arrays).

**Comment by He-Pin** [comment]
- [ ] She just share the number

**Comment by Glavo** [comment]
Test environment:

```
> neofetch
            .-/+oossssoo+/-.               glavo@minecraft-server
        `:+ssssssssssssssssss+:`           ----------------------
      -+ssssssssssssssssssyyssss+-         OS: Ubuntu 22.04.3 LTS x86_64
    .ossssssssssssssssssdMMMNysssso.       Kernel: 6.2.0-37-generic
   /ssssssssssshdmmNNmmyNMMMMhssssss/      Uptime: 53 days, 10 hours, 25 mins
  +ssssssssshmydMMMMMMMNddddyssssssss+     Packages: 2296 (dpkg), 13 (snap)
 /sssssssshNMMMyhhyyyyhmNMMMNhssssssss/    Shell: fish 3.3.1
.ssssssssdMMMNhsssssssssshNMMMdssssssss.   Terminal: /dev/pts/5
+sssshhhyNMMNyssssssssssssyNMMMysssssss+   CPU: AMD Ryzen 7 5800X (16) @ 4.600GHz
ossyNMMMNyMMhsssssssssssssshmmmhssssssso   GPU: NVIDIA GeForce GT 710
ossyNMMMNyMMhsssssssssssssshmmmhssssssso   Memory: 12174MiB / 31995MiB
+sssshhhyNMMNyssssssssssssyNMMMysssssss+
.ssssssssdMMMNhsssssssssshNMMMdssssssss.
 /sssssssshNMMMyhhyyyyhdNMMMNhssssssss/
  +sssssssssdmydMMMMMMMMddddyssssssss+
   /ssssssssssshdmNNNNmyNMMMMhssssss/
    .ossssssssssssssssssdMMMNysssso.
      -+sssssssssssssssssyyyssss+-
        `:+ssssssssssssssssss+:`
            .-/+oossssoo+/-.

> java -version
openjdk version "21.0.1" 2023-10-17 LTS
OpenJDK Runtime Environment (build 21.0.1+12-LTS)
OpenJDK 64-Bit Server VM (build 21.0.1+12-LTS, mixed mode, sharing)
```

Benchmark:
```java
import org.openjdk.jmh.annotations.*;

import java.util.Arrays;
import java.util.concurrent.TimeUnit;

@Warmup(iterations = 5, time = 3)
@Measurement(iterations = 5, time = 3)
@Fork(value = 1, jvmArgsAppend = {"-XX:+UseG1GC", "-Xms8g", "-Xmx8g"})
@BenchmarkMode(Mode.Throughput)
@OutputTimeUnit(TimeUnit.MILLISECONDS)
@State(Scope.Benchmark)
public class ArrayCopy {
    @Param({"1", "4", "16", "64", "256", "1024", "4096", "16384"})
    public int length;

    private Object[] array;

    @Setup
    public void setup() {
        array = new Object[length];
        for (int i = 0; i < length; i++) {
            array[i] = new Object();
        }
    }

    @Benchmark
    public Object[] arraycopy() {
        Object[] res = new Object[length * 2];
        System.arraycopy(array, 0, res, 0, length);
        return res;
    }

    @Benchmark
    public Object[] arraysCopyOf() {
        return Arrays.copyOf(array, length * 2);
    }
}
```

Results:
```
Benchmark               (length)   Mode  Cnt       Score      Error   Units
ArrayCopy.arraycopy            1  thrpt    5  248520.918 ±  312.114  ops/ms
ArrayCopy.arraycopy            4  thrpt    5  201618.928 ± 3571.060  ops/ms
ArrayCopy.arraycopy           16  thrpt    5  119598.976 ±  196.054  ops/ms
ArrayCopy.arraycopy           64  thrpt    5   29791.069 ±  243.624  ops/ms
ArrayCopy.arraycopy          256  thrpt    5    6436.471 ±    5.369  ops/ms
ArrayCopy.arraycopy         1024  thrpt    5    1824.224 ±  114.118  ops/ms
ArrayCopy.arraycopy         4096  thrpt    5     463.520 ±    0.927  ops/ms
ArrayCopy.arraycopy        16384  thrpt    5     107.259 ±    0.680  ops/ms
ArrayCopy.arraysCopyOf         1  thrpt    5  214251.970 ±  348.939  ops/ms
ArrayCopy.arraysCopyOf         4  thrpt    5  185852.802 ±  577.851  ops/ms
ArrayCopy.arraysCopyOf        16  thrpt    5  121371.523 ±  392.918  ops/ms
ArrayCopy.arraysCopyOf        64  thrpt    5   28541.141 ±   24.515  ops/ms
ArrayCopy.arraysCopyOf       256  thrpt    5    6424.406 ±    6.531  ops/ms
ArrayCopy.arraysCopyOf      1024  thrpt    5    1835.908 ±    2.270  ops/ms
ArrayCopy.arraysCopyOf      4096  thrpt    5     465.621 ±    0.326  ops/ms
ArrayCopy.arraysCopyOf     16384  thrpt    5     107.504 ±    0.122  ops/ms
```

**Comment by He-Pin** [comment]
image

File: core/shared/src/main/scala/cats/effect/unsafe/WeakBag.scala [hunk 2]
```diff
@@ -60,11 +59,8 @@ private final class WeakBag[A <: AnyRef] {
 
       // There are no expired entries, and the table has no leftover capacity.
       // The data structure needs to grow in this case.
-      val oldTable = table
-      val newCap = if (cap == MaxSizePow2) Int.MaxValue else cap << 1
-      val newTable = new Array[Entry[A]](newCap)
-      System.arraycopy(oldTable, 0, newTable, 0, idx)
-      table = newTable
+      val newCap = if (cap == MaxSizePow2) Int.MaxValue else 2 * cap
+      table = Arrays.copyOf(table, newCap)
```
Comment (He-Pin) [inline core/shared/src/main/scala/cats/effect/unsafe/WeakBag.scala:63]: image

Comment (He-Pin) [reply core/shared/src/main/scala/cats/effect/unsafe/WeakBag.scala:63]: No annotation

File: core/jvm-native/src/main/scala/cats/effect/ArrayStack.scala [hunk 3]
```diff
@@ -70,9 +72,7 @@ private final class ArrayStack[A <: AnyRef](
   private[this] def checkAndGrow(): Unit =
     if (index >= buffer.length) {
       val len = buffer.length
-      val buffer2 = new Array[AnyRef](len * 2)
-      System.arraycopy(buffer, 0, buffer2, 0, len)
-      buffer = buffer2
+      buffer = Arrays.copyOf(buffer, len * 2)
```
Comment (He-Pin) [inline core/jvm-native/src/main/scala/cats/effect/ArrayStack.scala:75]: I see, it's an AnyRef array!

**Review by He-Pin — APPROVED** [review]
lgtm

File: core/jvm-native/src/main/scala/cats/effect/ArrayStack.scala (hunk 3 above)
Comment (He-Pin) [reply core/jvm-native/src/main/scala/cats/effect/ArrayStack.scala:75]: image

**Comment by djspiewak** [comment]
So if I interpret your benchmark results correctly, @He-Pin, we should expect a very small performance improvement from this (for anything bounded by this codepath)?

**Comment by He-Pin** [comment]
No, I asked on JVM group, this will only help on large bytes, which is for SIMD.

**Comment by djspiewak** [comment]
Gotcha, so then we should probably just stick with the existing formulation here?

**Comment by He-Pin** [comment]
@djspiewak It works better on large bytes array, I have no idea.

**Comment by djspiewak** [comment]
Let's close this for now then. If large byte arrays are the sweet spot then we might see some benefits in `ByteStack` in pathological scenarios, but we've previously shown how `ByteStack` isn't the performance bottleneck for anything so it doesn't seem worthwhile to me.

**Pull request closed without merging**